### PR TITLE
Adding unit test for work item types

### PR DESCRIPTION
**Pull request by ldimaggi**
This PR replaces:

https://github.com/almighty/almighty-ui/pull/409

Thx!

### CODE REVIEW

**Comment by joshuawilson** [comment]
[test]

**Comment by joshuawilson** [comment]
[test]

File: src/app/work-item/work-item.service.ts [hunk 1]
```diff
@@ -57,6 +57,27 @@ export class WorkItemService {
       .catch(this.handleError);
   }
 
+ getWorkItemTypes(): Promise<any[]> {
+    let url = this.workItemTypeUrl + '.2';
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: This end point is not valid.

File: src/app/work-item/work-item.service.ts [hunk 2]
```diff
@@ -57,6 +57,27 @@ export class WorkItemService {
...
+      url = this.workItemTypeUrl;
+    }
+    return this.http
+      .get(url, { headers: this.headers })
+      .toPromise()
+      .then(response => {
+        if (process.env.ENV == 'inmemory') {
+          // Exclusively for in memory testing
+          this.workItemTypes = response.json().data as WorkItemType[];
+          this.workItemTypes = this.workItemTypes.reverse();
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Why would you reverse the work item types?

File: src/app/work-item/work-item.service.ts [hunk 3]
```diff
@@ -57,6 +57,27 @@ export class WorkItemService {
...
+        } else {
+          this.workItemTypes = response.json().data as WorkItemType[];
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: We don't have JSON API response format for **workitemtypes** api call. So, this will fail.

File: src/app/work-item/work-item.service.ts [hunk 4]
```diff
@@ -28,7 +28,7 @@ export class WorkItemService {
       this.headers.set('Authorization', 'Bearer ' + this.auth.getToken());
     }
     logger.log('WorkItemService running in ' + process.env.ENV + ' mode.');
-    logger.log('WorkItemService using url ' + this.workItemUrl);
+    logger.log('WorkItemService using urls ' + this.workItemUrl + ', ' + this.workItemTypeUrl);
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: I'd make different logs for showing URLs on different purpose.

File: src/app/work-item/work-item.service.spec.ts [hunk 5]
```diff
@@ -16,6 +16,7 @@ import { DropdownOption } from '../shared-component/dropdown/dropdown-option';
 import { Logger } from '../shared/logger.service';
 
 import { WorkItem } from './work-item';
+import { WorkItemType } from './work-item-type';
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.spec.ts]: Where is this file coming from?

**Comment by ldimaggi** [comment]
Changes have been incorporated - branch was rebased - it should be OK now

**Comment by joshuawilson** [comment]
[test]

**Comment by joshuawilson** [comment]
[test]

**Comment by joshuawilson** [comment]
[test]

File: src/app/work-item/work-item.service.ts [hunk 6]
```diff
@@ -57,6 +57,24 @@ export class WorkItemService {
...
   }
 
+ getWorkItemTypes(): Promise<any[]> {
+    this.logger.log('WorkItemService using url ' + this.workItemTypeUrl);
+    let url = this.workItemTypeUrl;
+    return this.http
+      .get(url, { headers: this.headers })
+      .toPromise()
+      .then(response => {
+        if (process.env.ENV == 'inmemory') {
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: Why was this broken into the longer "if" statement version? I appears to be an unnecessary change.

File: src/app/work-item/work-item.service.ts [hunk 7]
```diff
@@ -57,6 +57,24 @@ export class WorkItemService {
       .catch(this.handleError);
   }
 
+ getWorkItemTypes(): Promise<any[]> {
+    this.logger.log('WorkItemService using url ' + this.workItemTypeUrl);
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: Why is this being logged?

File: src/app/work-item/work-item.service.ts [hunk 8]
```diff
@@ -82,23 +100,6 @@ export class WorkItemService {
     return Promise.resolve(this.workItems);
   }
 
-  getWorkItemTypes(): Promise<any[]> {
-    if (this.workItemTypes.length) {
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: why was this check removed?

File: src/app/work-item/work-item.service.ts (hunk 7 above)
Comment (ldimaggi) [inline src/app/work-item/work-item.service.ts]: Sudipta requested an additional log message when a URL other than that used for the work items was accessed.

File: src/app/work-item/work-item.service.ts [hunk 9]
```diff
@@ -82,23 +107,6 @@ export class WorkItemService {
     return Promise.resolve(this.workItems);
   }
 
-  getWorkItemTypes(): Promise<any[]> {
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Please restore this code from 85 to 101

File: src/app/work-item/work-item.service.ts [hunk 10]
```diff
@@ -57,6 +57,31 @@ export class WorkItemService {
       .catch(this.handleError);
   }
 
+ getWorkItemTypes(): Promise<any[]> {
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Please remove this code from 60 to 84

**Comment by sanbornsen** [comment]
This is all good now.
